pkg/oidcclient/login.go

```diff
@@ -822,7 +822,7 @@ func (h *handlerState) handleRefresh(ctx context.Context, refreshToken *oidctype
 
 	// The spec is not 100% clear about whether an ID token from the refresh flow should include a nonce, and at least
 	// some providers do not include one, so we skip the nonce validation here (but not other validations).
-	return upstreamOIDCIdentityProvider.ValidateToken(ctx, refreshed, "")
+	return upstreamOIDCIdentityProvider.ValidateTokenAndMergeWithUserInfo(ctx, refreshed, "", true)
```

Comment:
Interesting. I can't recall why it was originally done this way. Perhaps @cfryanr remembers some of the context? We still claim to support oidc providers other than the supervisor with the cli (although in practice I'm not sure anyone uses it this way).

Reply:
I don't think we noticed this when we discussed this code before the holidays, Margo. I think this PR didn't add those behaviors, but only renamed the function to make it more obvious what it was already doing, right? It's nice that the more clear function name helped point out a small performance concern for the CLI here.

Historically, I suspect fetching the keyset was due to how the original author decided to interpret https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation and because that's what the underlying go-oidc library does by default. The userinfo stuff was added later, and maybe that enhancement didn't fully consider which code was shared with the CLI. It does seem unnecessary for the CLI to fetch userinfo in hindsight.

In practice the userinfo check is skipped here anyway when the server is a Supervisor, since the Supervisor doesn't advertise a userinfo endpoint.

The CLI could probably do a simplified ID token validation which skips any possible userinfo calls and takes advantage of this line from the spec to also skip fetching the issuer keys while remaining spec-compliant:

> If the ID Token is received via direct communication between the Client and the Token Endpoint (which it is in this flow), the TLS server validation MAY be used to validate the issuer in place of checking the token signature.

I wonder if we could trick the go-oidc library into skipping signature verification, maybe by implementing a no-op version of a KeySet. This would only be safe if we strictly prohibit non-TLS OIDC providers in the CLI.

Reply:
AFAIK, we always require TLS for OIDC issuers as issuers URLs are required to be `https` per the spec, and we provide no option to skip TLS verification (without TLS the key set cannot be trusted anyway).  Thus to me, `handlerState.handleRefresh` should simply call `upstreamOIDCIdentityProvider.PerformRefresh` and that is all (i.e. drop the call to `upstreamOIDCIdentityProvider.ValidateTokenAndMergeWithUserInfo` altogether).  I do not see any reason to do any verification on the returned tokens as there are no token substitution attacks on your local CLI.

@margocrawf @cfryanr WDYT?

Reply:
Resolved via making a separate issue to address this.